### PR TITLE
update cell UI directly after converting cell type (#5053)

### LINKED ISSUE #5053: Cell type not updated in UI when converting between Markdown and code
**Application or Package Used**

nteract desktop v 0.22.4 

**Describe the bug**

The cell type is not updated as expected in the user interface when converting a cell between from markdown and code cell or vice-versa.

**To Reproduce**

1. open any notebook
1. create a markdown cell
1. write some markdown in it, eg. "# header1"
1. use the cell menu to convert the cell into a code cell
1. the cell still looks like a markdown cell (e.g incorrect syntax highlighting), as if it wasn't converted.
1. you need to activate another cell before the UI is updated. Or execute the cell with shift-enter, which updates it, because it automatically activates the next cell at the same time. 

**Expected behavior**

When you click the menu option to convert the cell, the UI should update right away to reflect that change.

E.g this is currently a _Markdown_ cell, even though it looks like a code cell. Note that the "convert to Code Cell" option is in the menu, because internal state is still a markdown cell, even though the UI is not.
image

**Desktop (please complete the following information):**

- OS: Ubuntu 19.10
- Version: 0.22.4

### PR DESCRIPTION
**Pull request by jarmokivekas**
This PR proposes a fix for  #5053 

Using _Convert to Markdown  Cell_ and _Convert to Code Cell_ work as in earlier versions, clicking the menu instantly updates the cell as well.

### CODE REVIEW

**Comment by jarmokivekas** [comment]
Testing manually on the desktop using `yarn spawn` this works fine like this. Don't know how much you do UI testing automatically, but i have 57 / 84 test suites failing in `yarn test` on current master, soo... They're not super useful right now. Probably some devel environment misconfiguration.

**Comment by jdetle** [comment]
Hey @jarmokivekas, can you provide a notebook that previously had this error and is fixed by your PR? 

> Don't know how much you do UI testing automatically, but i have 57 / 84 test suites failing in yarn test on current master, soo... They're not super useful right now. Probably some devel environment misconfiguration.

With Jest this can often come from an older version of node, can you ensure you're on node 12 or greater?

**Comment by jarmokivekas** [comment]
My node version is v12.16.2

As far as specific notebooks go, on current master any notebook has this bug. If I create a new notebook with one empty cell, it won't convert in the UI until it is un-selected once.

**Comment by jarmokivekas** [comment]
The tests are all pretty much failing due to import errors: https://gist.github.com/jarmokivekas/7b4b6696902bdebc1974fb27ee1eedf6

I'm running `yarn test` from the repo root after having ran `yarn`

File: packages/stateful-components/src/cells/cell.tsx [hunk 1]
```diff
@@ -19,7 +19,15 @@ type Props = ComponentProps & StateProps;
 
 export class Cell extends React.Component<Props> {
   shouldComponentUpdate(nextProps: Props): boolean {
-    return nextProps.selected !== this.props.selected;
+    // Update cell directly after converting between markdown and code
+    if (this.props.cell.cell_type !== nextProps.cell.cell_type) {
```
Comment (jdetle) [inline packages/stateful-components/src/cells/cell.tsx]: `
```suggestion
    if (this.props.cell?.cell_type !== nextProps.cell?.cell_type) {

```

File: packages/stateful-components/src/cells/cell.tsx [hunk 2]
```diff
@@ -19,7 +19,15 @@ type Props = ComponentProps & StateProps;
 
 export class Cell extends React.Component<Props> {
   shouldComponentUpdate(nextProps: Props): boolean {
-    return nextProps.selected !== this.props.selected;
+    // Update cell directly after converting between markdown and code
+    if (this.props.cell.cell_type !== nextProps.cell.cell_type) {
+      return true;
+    }
+    // Update cell when selecting or de-selecting it
+    if (nextProps.selected !== this.props.selected) {
```
Comment (jdetle) [inline packages/stateful-components/src/cells/cell.tsx]:
```suggestion
    if (nextProps?.selected !== this.props?.selected) {

```

**Comment by jarmokivekas** [comment]
I accepted the review change suggestions.

I'm happy if you are @jdetle

**Comment by jdetle** [comment]
Thanks for catching this @jarmokivekas

**Pull request merged**